src/orders/components/OrderGrantRefundPage/reducer.ts

```diff
@@ -0,0 +1,127 @@
+import { OrderDetailsGrantRefundFragment } from "@saleor/graphql";
+
+export interface ReducerOrderLine {
+  selectedQuantity: number;
+  availableQuantity: number;
+  unitPrice: number;
+}
+
+export interface GrantRefundState {
+  lines: Map<string, ReducerOrderLine>;
+  refundShipping: boolean;
+}
+
+export type GrantRefundLineKeyValue = [string, ReducerOrderLine];
+
+export type GrantRefundAction =
+  | {
+      type: "setQuantity";
+      lineId: string;
+      amount: number;
+    }
+  | {
+      type: "setMaxQuantity";
+      lineIds: string[];
+    }
+  | {
+      type: "initState";
+      state: GrantRefundState;
+    }
+  | {
+      type: "toggleRefundShipping";
+    };
```

Comment:
love it, so that's actually a discrimination union pattern you can also add an exhaustiveness check (optional)